### PR TITLE
Small cleanup of ComputeAtRootDomainMap

### LINKED ISSUE #961: index_select failed on broadcasted inputs.
Calling index_select on broadcast dimension gives Runtime Error.
```
RuntimeError: producer_id->isBroadcast() INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/root_domain_map.cpp":887, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues.
Exception raised from setMaybeMapped at /opt/pytorch/nvfuser/csrc/root_domain_map.cpp:887 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, char const*) + 0xe3 (0x7fe723f9fe99 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #1: nvfuser::ComputeAtRootDomainMapBuilder::setMaybeMapped(nvfuser::TensorDomain const*, nvfuser::IterDomain const*, nvfuser::TensorDomain const*, nvfuser::IterDomain const*) + 0x3e4 (0x7fe72440c964 in /usr/
local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #2: nvfuser::ComputeAtRootDomainMapBuilder::mapPointwiseOrReductionOp(nvfuser::Expr*) + 0x382 (0x7fe72440cd22 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #3: nvfuser::ComputeAtRootDomainMapBuilder::dispatch(nvfuser::Expr*) + 0x7a (0x7fe72440e5ea in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #4: nvfuser::BackwardVisitor::traverseTo(nvfuser::Fusion*, std::vector<nvfuser::Val*, std::allocator<nvfuser::Val*> > const&, bool) + 0xd73 (0x7fe724225733 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #5: nvfuser::ComputeAtRootDomainMapBuilder::ComputeAtRootDomainMapBuilder(nvfuser::ComputeAtRootDomainMap&, bool) + 0x207 (0x7fe724406187 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #6: nvfuser::ComputeAtRootDomainMap::build(bool) + 0x2bd (0x7fe72440671d in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #7: nvfuser::MaxPosCalculator::buildUnmappableDims(bool) + 0x254 (0x7fe72401b394 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #8: nvfuser::MaxPosCalculator::MaxPosCalculator(std::unordered_set<nvfuser::IterDomain*, std::hash<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<nvfuser::IterDomain*> >, bool) + 0xcf (0x7fe72401bcaf in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #9: nvfuser::inlineAllAt(nvfuser::TensorView*, long, bool, std::unordered_set<nvfuser::IterDomain*, std::hash<nvfuser::IterDomain*>, std::equal_to<nvfuser::IterDomain*>, std::allocator<nvfuser::IterDomain*> > const&) + 0x99 (0x7fe72401c299 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #10: nvfuser::schedulePointwise(nvfuser::Fusion*, nvfuser::PointwiseParams const&) + 0x177c (0x7fe7244242dc in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #11: nvfuser::PointWiseScheduler::schedule(nvfuser::Fusion*) + 0xc6 (0x7fe724425476 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #12: nvfuser::FusionKernelRuntime::compileKernel(nvfuser::KernelArgumentHolder const&, nvfuser::SegmentedGroup*) + 0x13e (0x7fe724234e1e in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #13: nvfuser::FusionKernelRuntime::compileFusionParallel(nvfuser::KernelArgumentHolder) + 0x3e1 (0x7fe724239dc1 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #14: nvfuser::FusionExecutorCache::runFusionWithInputs(c10::ArrayRef<c10::IValue> const&, std::optional<nvfuser::PrimDataType>, std::optional<signed char>) + 0xa45 (0x7fe724243695 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
frame #15: nvfuser::python_frontend::FusionDefinition::execute(c10::ArrayRef<c10::IValue> const&, bool, bool, std::optional<signed char>) const + 0x331 (0x7fe724544f01 in /usr/local/lib/python3.10/site-packages/torch/lib/libnvfuser_codegen.so)
```

Here's the repro script:
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[1], contiguity=[None], dtype=DataType.Int, is_cpu=False)
    T1 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.ops.index_select(T1, T0, dim=0)
    fd.add_output(T2)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randint(0, 10, (1,), dtype=torch.int64, device='cuda:0').as_strided((1,), (1,)),
    torch.randn((9216000,), dtype=torch.float32, device='cuda:0').as_strided((32000, 288), (288, 1)),
]
fd.execute(inputs)
```

### PR DESCRIPTION
**Pull request by naoyam**
This also fixes #961

Eventually ComputeAtRootDomainMap should also be replaced with the Index Graph

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by carmocca** [comment]
I verified that the original error is gone after building nvFuser from this branch. Thank you!

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

File: csrc/root_domain_map.cpp [hunk 1]
```diff
@@ -942,50 +948,39 @@ void ComputeAtRootDomainMapBuilder::dispatch(Expr* e) {
...
-            " is found");
-        for (auto out : e->outputs()) {
-          auto out_tv = out->as<TensorView>();
-          auto out_td = out_tv->domain();
-          auto out_root = out_td->root();
-          setMaybeMapped(in_td, in_root[it], out_td, out_root[it]);
-        }
-      } else {
-        setMaybeMapped(in_td, in_root[it], out_td, out_root[it]);
+  for (auto producer_tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
```
Comment (naoyam) [inline csrc/root_domain_map.cpp:973]: Don't remember why this wasn't done this way, but just using `PairwiseRootDomainMap` should work fine and simpler.

File: csrc/root_domain_map.h [hunk 2]
```diff
@@ -438,58 +438,64 @@ class ComputeAtRootDomainMapBuilder : private BackwardVisitor {
...
 
   void handle(TernaryOp* top) override {
-    mapPointwiseOrReductionOp(top);
+    mapPointwiseLikeOp(top);
   }
 
   void handle(RNGOp* top) override;
 
-  void handle(IndexSelectOp* top) override {
-    mapPointwiseOrReductionOp(top);
```
Comment (naoyam) [inline csrc/root_domain_map.h:462]: This wasn't correct before as the previous version of `mapPointwiseOrReductionOp` did not consider indexed IDs, which are now properly handled by `PairwiseRootDomainMap`.

File: csrc/root_domain_map.cpp [hunk 3]
```diff
@@ -942,50 +948,39 @@ void ComputeAtRootDomainMapBuilder::dispatch(Expr* e) {
...
+    for (auto consumer_tv :
+         ir_utils::filterByType<TensorView>(expr->outputs())) {
+      for (const auto& mapping : PairwiseRootDomainMap(producer_tv, consumer_tv)
+                                     .mapBroadcast(true)
+                                     .mapProducerToConsumer()) {
+        setMaybeMapped(
```
Comment (naoyam) [inline csrc/root_domain_map.cpp:979]: This analysis gathers all mapping information between producer and consumer domains where inlining should be allowed. For example, a broadcast producer domain should be allowed to be inlined to a non-broadcast consumer domain. However, it should not be allowed to inline an indexed producer domain into the consumer domain as the actual index of the producer domain is unknown.

File: csrc/root_domain_map.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/root_domain_map.h:462]: Can we also use this `mapPointwiseLikeOp` for broadcast and squeeze? Looks like every op should be handled by this function.

Comment (zasdfgbnm) [reply csrc/root_domain_map.h:462]: I think we will need to update `new_broadcast_domains_` and `removed_broadcast_domains_` besides doing `mapPointwiseLikeOp`

Comment (naoyam) [reply csrc/root_domain_map.h:462]: Yes, the broadcast analysis also needs to be done.

I didn't want to spend too much effort to clean up this class as this should eventually be replaced with IDGraph. The cleanup I did in this PR was part of the fix for #961.

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**